### PR TITLE
8305083: Remove finalize() from test/hotspot/jtreg/vmTestbase/nsk/share/ and /jpda that are used in serviceability/dcmd/framework tests

### PR DESCRIPTION
**Pull request by afshin-zafari**
This PR is continuation of https://github.com/openjdk/jdk/pull/13420 which was far behind the master.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305083](https://bugs.openjdk.org/browse/JDK-8305083): Remove finalize() from test/hotspot/jtreg/vmTestbase/nsk/share/ and /jpda that are used in serviceability/dcmd/framework tests

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [63e342f6](https://git.openjdk.org/jdk/pull/13884/files/63e342f6cbf1ebcbb6a103c776f9c654673815c3)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [63e342f6](https://git.openjdk.org/jdk/pull/13884/files/63e342f6cbf1ebcbb6a103c776f9c654673815c3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13884/head:pull/13884` \
`$ git checkout pull/13884`

Update a local copy of the PR: \
`$ git checkout pull/13884` \
`$ git pull https://git.openjdk.org/jdk.git pull/13884/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13884`

View PR using the GUI difftool: \
`$ git pr show -t 13884`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13884.diff">https://git.openjdk.org/jdk/pull/13884.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13884#issuecomment-1539768258)

### CODE REVIEW

**Comment by afshin-zafari** [comment]
Brief reminder from original PR:
- `Finalizable` interface has `default` methods for `cleanup` and `registerCleanup`.
- `FinalizableObject` implements that interface.
- All the classes implementing the `Finalizable` interface just call the default methods. No need to override.
- When needed specific cleanup actions, the `cleanup()` method is overridden.

**Comment by afshin-zafari** [comment]
reply to @dholmes-ora comment:
> I think a better design for the classes that can't extend FinalizableObject would be for them to contain a FinalizableObject, the cleanup action for which would cleanup the host object. That could allow the removal of the Finalizable interface and simplify the general usage patterns. But that may be going too far for this particular PR ... ?

We still need the `Finalizable` interface to call the `cleanup` of the host object:
```java
class FinalizableObject implements Finalizable { 
 public void registerCleanupForObject(Finalizable obj) { 
  Cleaner.create().register(obj, obj->cleanup()); 
}
```

File: test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java [hunk 1]
```diff
@@ -166,12 +164,10 @@ protected void kill() {
     }
 
     /**
-     * Finalize mirror by invoking <code>close()</code>.
+     * This method is called at finalization and calls <code>kill()</code>.
      *
-     * @throws Throwable if any throwable exception is thrown during finalization
      */
-    protected void finalize() throws Throwable {
+    public void cleanup() {
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java:171]: Fixed.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java [hunk 2]
```diff
@@ -408,19 +407,17 @@ public synchronized void close() {
      *
      * @see #close()
      */
-    protected void finalize() throws Throwable {
+    public void cleanup() {
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java:410]: Fixed.

**Review by afshin-zafari — COMMENTED** [review]
I fixed Override and removed the code that came after merging with master.

**Comment by dholmes-ora** [comment]
> We still need the Finalizable interface to call the cleanup of the host object:

No you would replace any occurrence of `Finalizable` with `FinalizeableObject`

File: test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java [hunk 3]
```diff
@@ -23,28 +23,18 @@
 
 package nsk.share;
 
+
 /**
- * This class is an simple exalmple of finalizable object, that implements interface
- * <code>Finalizable</code> and invokes standard <code>finalize()</code> method
- * as finalization.
+ * This class is an simple example of finalizable object, that
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java:28]: s/an/a/

**Review by dholmes-ora — APPROVED** [review]
I think this will have to do for now - it is hard to make small adjustments to a lot of these old tests as they just cry out to be rewritten in a cleaner way. But we can look at that in the follow up RFE.

Thanks.

**Comment by afshin-zafari** [comment]
Thank you @coleenp and @dholmes-ora for your comments.
/integrate

**Pull request closed without merging**